In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

#load data

In [ ]:
data_train = "https://docs.google.com/spreadsheets/d/e/2PACX-1vSG8gRBml0Y1N8nxlFGkUyG4bp0gFllEobzC7nd1fM6TjHdWFHTn_GPEKcQWb_SEcuDaEDflX_14hZ2/pub?gid=1270430478&single=true&output=csv"

train = pd.read_csv(data_train)
train

id  DC024  DC025  DC205  DC206  DC207  DC208  DC209  DC210  DC211  \
0          1     11      2   16.0    1.0    0.0    1.0    0.0    0.0    0.0   
1          2     11      2   31.0    1.0    0.0    1.0    0.0    0.0    1.0   
2          3     11      2   12.0    1.0    0.0    1.0    1.0    1.0    1.0   
3          6     11      2   12.0    1.0    0.0    0.0    0.0    0.0    1.0   
4          7     11      2   12.0    1.0    0.0    1.0    1.0    0.0    0.0   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
35968  47958     94      1   17.0    1.0    0.0    0.0    0.0    0.0    0.0   
35969  47959     94      1   12.0    1.0    0.0    1.0    1.0    0.0    0.0   
35970  47961     94      1   12.0    1.0    0.0    1.0    1.0    0.0    1.0   
35971  47962     94      1   12.0    1.0    0.0    1.0    1.0    0.0    1.0   
35972  47963     94      1   12.0    1.0    0.0    1.0    1.0    0.0    1.0   

       ...  DC237f  DC241  DC242  DC244  DC246  DC252  DC270a  DC109  DC142a  \
0      ...     0.0    1.0    0.0    0.0    0.0    2.0       1   12.0    60.0   
1      ...     0.0    1.0    0.0    1.0    1.0    1.0       2   31.0    70.0   
2      ...     0.0    1.0    0.0    1.0    1.0    0.0       4   11.0    75.0   
3      ...     0.0    2.0    NaN    1.0    0.0    1.0       1   11.0    85.0   
4      ...     0.0    1.0    0.0    0.0    1.0    0.0       2   11.0    65.0   
...    ...     ...    ...    ...    ...    ...    ...     ...    ...     ...   
35968  ...     0.0    3.0    NaN    0.0    0.0    0.0       1   21.0     9.0   
35969  ...     0.0    1.0    1.0    0.0    0.0    0.0       2   11.0    12.0   
35970  ...     0.0    1.0    1.0    0.0    1.0    1.0       1   11.0    48.0   
35971  ...     0.0    1.0    1.0    0.0    0.0    0.0       2   11.0    39.0   
35972  ...     0.0    1.0    0.0    0.0    0.0    0.0       3   11.0    40.0   

             DC201  
0      Layak Minum  
1      Layak Minum  
2      Layak Minum  
3      Layak Minum  
4      Layak Minum  
...            ...  
35968  Layak Minum  
35969  Layak Minum  
35970  Layak Minum  
35971  Layak Minum  
35972  Layak Minum  

[35973 rows x 40 columns]

#data understanding

In [ ]:
train.head()

id  DC024  DC025  DC205  DC206  DC207  DC208  DC209  DC210  DC211  ...  \
0   1     11      2   16.0    1.0    0.0    1.0    0.0    0.0    0.0  ...   
1   2     11      2   31.0    1.0    0.0    1.0    0.0    0.0    1.0  ...   
2   3     11      2   12.0    1.0    0.0    1.0    1.0    1.0    1.0  ...   
3   6     11      2   12.0    1.0    0.0    0.0    0.0    0.0    1.0  ...   
4   7     11      2   12.0    1.0    0.0    1.0    1.0    0.0    0.0  ...   

   DC237f  DC241  DC242  DC244  DC246  DC252  DC270a  DC109  DC142a  \
0     0.0    1.0    0.0    0.0    0.0    2.0       1   12.0    60.0   
1     0.0    1.0    0.0    1.0    1.0    1.0       2   31.0    70.0   
2     0.0    1.0    0.0    1.0    1.0    0.0       4   11.0    75.0   
3     0.0    2.0    NaN    1.0    0.0    1.0       1   11.0    85.0   
4     0.0    1.0    0.0    0.0    1.0    0.0       2   11.0    65.0   

         DC201  
0  Layak Minum  
1  Layak Minum  
2  Layak Minum  
3  Layak Minum  
4  Layak Minum  

[5 rows x 40 columns]

In [ ]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35973 entries, 0 to 35972
Data columns (total 40 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   id      35973 non-null  int64  
 1   DC024   35973 non-null  int64  
 2   DC025   35973 non-null  int64  
 3   DC205   35962 non-null  float64
 4   DC206   35952 non-null  float64
 5   DC207   35929 non-null  float64
 6   DC208   35960 non-null  float64
 7   DC209   35955 non-null  float64
 8   DC210   35932 non-null  float64
 9   DC211   35949 non-null  float64
 10  DC212   35949 non-null  float64
 11  DC213   35896 non-null  float64
 12  DC214   35964 non-null  float64
 13  DC215   35959 non-null  float64
 14  DC216   35934 non-null  float64
 15  DC217   35973 non-null  int64  
 16  DC219   35973 non-null  int64  
 17  DC220   35969 non-null  float64
 18  DC226   35967 non-null  float64
 19  DC230a  35962 non-null  float64
 20  DC230b  32964 non-null  float64
 21  DC232   32949 non-null  float64
 22

In [ ]:
train.isnull().sum()

id            0
DC024         0
DC025         0
DC205        11
DC206        21
DC207        44
DC208        13
DC209        18
DC210        41
DC211        24
DC212        24
DC213        77
DC214         9
DC215        14
DC216        39
DC217         0
DC219         0
DC220         4
DC226         6
DC230a       11
DC230b     3009
DC232      3024
DC232b     3024
DC235     10446
DC237        21
DC237a       37
DC237b       37
DC237c       37
DC237d       37
DC237e       37
DC237f       37
DC241       612
DC242      5374
DC244         4
DC246         9
DC252        14
DC270a        0
DC109        11
DC142a       52
DC201         3
dtype: int64

In [ ]:
train1 = pd.DataFrame(train.isnull().sum().sort_values(ascending=False))
print(train1.to_markdown())

|        |     0 |
|:-------|------:|
| DC235  | 10446 |
| DC242  |  5374 |
| DC232  |  3024 |
| DC232b |  3024 |
| DC230b |  3009 |
| DC241  |   612 |
| DC213  |    77 |
| DC142a |    52 |
| DC207  |    44 |
| DC210  |    41 |
| DC216  |    39 |
| DC237d |    37 |
| DC237a |    37 |
| DC237b |    37 |
| DC237c |    37 |
| DC237f |    37 |
| DC237e |    37 |
| DC212  |    24 |
| DC211  |    24 |
| DC237  |    21 |
| DC206  |    21 |
| DC209  |    18 |
| DC252  |    14 |
| DC215  |    14 |
| DC208  |    13 |
| DC205  |    11 |
| DC109  |    11 |
| DC230a |    11 |
| DC246  |     9 |
| DC214  |     9 |
| DC226  |     6 |
| DC220  |     4 |
| DC244  |     4 |
| DC201  |     3 |
| DC024  |     0 |
| DC219  |     0 |
| DC270a |     0 |
| DC217  |     0 |
| DC025  |     0 |
| id     |     0 |


In [ ]:
# drop column
train = train.drop(['DC206','DC207','DC208','DC209','DC213','DC214','DC215','DC216','DC220','DC230a','DC230b','DC232','DC232b','DC235','DC241','DC242','DC244','DC246','DC252','DC142a'],axis=1)
train.isna().sum()

id         0
DC024      0
DC025      0
DC205     11
DC210     41
DC211     24
DC212     24
DC217      0
DC219      0
DC226      6
DC237     21
DC237a    37
DC237b    37
DC237c    37
DC237d    37
DC237e    37
DC237f    37
DC270a     0
DC109     11
DC201      3
dtype: int64

In [ ]:
# drop row
train = train.dropna(subset = ['DC201','DC205'])
train.isna().sum()

id         0
DC024      0
DC025      0
DC205      0
DC210     41
DC211     24
DC212     24
DC217      0
DC219      0
DC226      6
DC237     19
DC237a    35
DC237b    35
DC237c    35
DC237d    35
DC237e    35
DC237f    35
DC270a     0
DC109      0
DC201      0
dtype: int64

In [ ]:
# fill with mode
from sklearn.impute import SimpleImputer

imputer = SimpleImputer(strategy='most_frequent', missing_values=np.nan)
train[['DC210','DC211','DC212','DC226','DC237','DC237a','DC237b','DC237c','DC237d','DC237e','DC237f','DC109','DC226']] = imputer.fit_transform(train[['DC210','DC211','DC212','DC226','DC237','DC237a','DC237b','DC237c','DC237d','DC237e','DC237f','DC109','DC226']])

In [ ]:
train.isna().sum()

id        0
DC024     0
DC025     0
DC205     0
DC210     0
DC211     0
DC212     0
DC217     0
DC219     0
DC226     0
DC237     0
DC237a    0
DC237b    0
DC237c    0
DC237d    0
DC237e    0
DC237f    0
DC270a    0
DC109     0
DC201     0
dtype: int64

In [ ]:
train = train.drop('id', axis = 1)
train.isna().sum()

DC024     0
DC025     0
DC205     0
DC210     0
DC211     0
DC212     0
DC217     0
DC219     0
DC226     0
DC237     0
DC237a    0
DC237b    0
DC237c    0
DC237d    0
DC237e    0
DC237f    0
DC270a    0
DC109     0
DC201     0
dtype: int64

In [ ]:
train.head()

DC024  DC025  DC205  DC210  DC211  DC212  DC217  DC219  DC226  DC237  \
0     11      2   16.0    0.0    0.0    0.0      1      2    8.0    1.0   
1     11      2   31.0    0.0    1.0    0.0      4      1    2.0    1.0   
2     11      2   12.0    1.0    1.0    0.0      2      1    2.0    1.0   
3     11      2   12.0    0.0    1.0    0.0      4      1    8.0    1.0   
4     11      2   12.0    0.0    0.0    0.0      3      2    8.0    1.0   

   DC237a  DC237b  DC237c  DC237d  DC237e  DC237f  DC270a  DC109        DC201  
0     1.0     0.0     0.0     0.0     0.0     0.0       1   12.0  Layak Minum  
1     1.0     0.0     0.0     0.0     0.0     0.0       2   31.0  Layak Minum  
2     1.0     0.0     0.0     0.0     0.0     0.0       4   11.0  Layak Minum  
3     1.0     0.0     0.0     0.0     0.0     0.0       1   11.0  Layak Minum  
4     1.0     0.0     0.0     1.0     0.0     0.0       2   11.0  Layak Minum

In [ ]:
train['DC201'] = train.DC201.map({'Layak Minum': 1, 'Tidak Layak Minum': 0})

In [ ]:
feat_data = train.drop(columns =['DC201'])
feat_data.head()

DC024  DC025  DC205  DC210  DC211  DC212  DC217  DC219  DC226  DC237  \
0     11      2   16.0    0.0    0.0    0.0      1      2    8.0    1.0   
1     11      2   31.0    0.0    1.0    0.0      4      1    2.0    1.0   
2     11      2   12.0    1.0    1.0    0.0      2      1    2.0    1.0   
3     11      2   12.0    0.0    1.0    0.0      4      1    8.0    1.0   
4     11      2   12.0    0.0    0.0    0.0      3      2    8.0    1.0   

   DC237a  DC237b  DC237c  DC237d  DC237e  DC237f  DC270a  DC109  
0     1.0     0.0     0.0     0.0     0.0     0.0       1   12.0  
1     1.0     0.0     0.0     0.0     0.0     0.0       2   31.0  
2     1.0     0.0     0.0     0.0     0.0     0.0       4   11.0  
3     1.0     0.0     0.0     0.0     0.0     0.0       1   11.0  
4     1.0     0.0     0.0     1.0     0.0     0.0       2   11.0

In [ ]:
label_data = train['DC201']
label_data.head()

0    1
1    1
2    1
3    1
4    1
Name: DC201, dtype: int64

#Split data

In [ ]:
from sklearn.model_selection import train_test_split
xtrain, xtest, ytrain, ytest = train_test_split(feat_data, label_data, test_size=0.2, random_state=500)

#Standarization

In [ ]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
sc.fit(xtrain)
Xtrain_std = sc.transform(xtrain)
Xtest_std = sc.transform(xtest)


#DECISION TREE

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.tree import DecisionTreeClassifier
from sklearn import datasets
from sklearn import metrics

dt_cl = DecisionTreeClassifier (random_state =500)
#fit the model
dt_cl.fit (Xtrain_std, ytrain)

# Make the predictions
y_preddt_test = dt_cl.predict(Xtest_std)
y_preddt_train= dt_cl.predict(Xtrain_std)
 
# Measure the performance
print("Accuracytest score %.3f" %metrics.accuracy_score(ytest, y_preddt_test))
print("Accuracytrain score %.3f" %metrics.accuracy_score(ytrain, y_preddt_train))

Accuracytest score 0.880
Accuracytrain score 0.960


#SVM

In [ ]:
# Sklearn modules & classes
from sklearn.linear_model import Perceptron, LogisticRegression
from sklearn.svm import SVC

# Instantiate the Support Vector Classifier (SVC)
svc = SVC(C=1.0, random_state=1, kernel='linear')
 
# Fit the model
svc.fit(Xtrain_std, ytrain)

# Make the predictions
y_predict = svc.predict(Xtest_std)
y_predicttrain= svc.predict(Xtrain_std)
 
# Measure the performance
print("Accuracytest score %.3f" %metrics.accuracy_score(ytest, y_predict))
print("Accuracytrain score %.3f" %metrics.accuracy_score(ytrain, y_predicttrain))

Accuracytest score 0.905
Accuracytrain score 0.900


#NAIVE BAYES

In [ ]:
from sklearn.naive_bayes import GaussianNB
nb = GaussianNB()

# Fit the model
nb.fit (Xtrain_std, ytrain)

# Make the predictions
y_prednb= nb.predict(Xtest_std)
y_predtrainnb= nb.predict(Xtrain_std)
 
# Measure the performance
print("Accuracytest score %.3f" %metrics.accuracy_score(ytest, y_prednb))
print("Accuracytrain score %.3f" %metrics.accuracy_score(ytrain, y_predtrainnb))

Accuracytest score 0.663
Accuracytrain score 0.660


#Logistic regression

In [ ]:
from sklearn.linear_model import LogisticRegression

# Fit the model
logreg = LogisticRegression(solver='liblinear', random_state=0)
logreg.fit(feat_data, label_data)

# Make the predictions
y_predlr= logreg.predict(Xtest_std)
y_predtrainlr= logreg.predict(Xtrain_std)

# Measure the performance
print("Accuracytest score %.3f" %metrics.accuracy_score(ytest, y_predlr))
print("Accuracytrain score %.3f" %metrics.accuracy_score(ytrain, y_predtrainlr))

Accuracytest score 0.904
Accuracytrain score 0.899


#Random Forest

In [ ]:
# Instantiate the Support Vector Classifier (SVC)
from sklearn import ensemble
rf_cl = ensemble.RandomForestClassifier (n_estimators=100)

# Fit the model
rf_cl.fit (Xtrain_std, ytrain)

# Make the predictions
y_predrfcl= rf_cl.predict(Xtest_std)
y_predtrainrfcl= rf_cl.predict(Xtrain_std)
 
# Measure the performance
print("Accuracytest score %.3f" %metrics.accuracy_score(ytest, y_predrfcl))
print("Accuracytrain score %.3f" %metrics.accuracy_score(ytrain, y_predtrainrfcl))

Accuracytest score 0.900
Accuracytrain score 0.960


#Gradient Boosting

In [ ]:
from sklearn.metrics import classification_report
from sklearn.datasets import load_breast_cancer
from sklearn.ensemble import GradientBoostingClassifier

from sklearn.metrics.pairwise import gen_batches
gb= GradientBoostingClassifier(learning_rate=0.1)
#fit the model
gb.fit (Xtrain_std, ytrain)

# Make the predictions
y_predgb= gb.predict(Xtest_std)
y_predtraingb= gb.predict(Xtrain_std)
 
# Measure the performance
print("Accuracytest score %.3f" %metrics.accuracy_score(ytest, y_predgb))
print("Accuracytrain score %.3f" %metrics.accuracy_score(ytrain, y_predtraingb))

Accuracytest score 0.914
Accuracytrain score 0.910


#KNN

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

knn_cl = KNeighborsClassifier (n_neighbors=3)
#fit the model
knn_cl.fit (Xtrain_std, ytrain)
# Make the predictions
y_predknn_cl= knn_cl.predict(Xtest_std)
y_predtrainknn_cl= knn_cl.predict(Xtrain_std)
 
# Measure the performance
print("Accuracytest score %.3f" %metrics.accuracy_score(ytest, y_predknn_cl))
print("Accuracytrain score %.3f" %metrics.accuracy_score(ytrain, y_predtrainknn_cl))

Accuracytest score 0.893
Accuracytrain score 0.926


**Kesimpulan pemilihan model**

Dari beberapa model diatas, model terbaik yang digunakan untuk melakukan prediksi permasalahan diatas adalah model dengan akurasi terbaik yaitu Gradient Boosting karena perbedaan akurasi prediksi data test dengan data train hanya selisih 0.04 dan akurasi masing-masing dari prediksi data test dan data train adalah 0.914 dan 0.910. Berikut ini ditampilkan hasil lebih lanjut mengenai confusion matrix untuk model Gradient Boosting beserta perbandingan nilai prediksi dan aktualnya.

In [ ]:
from sklearn.metrics import confusion_matrix
confusion_matrix(ytest, y_predgb)

array([[ 135,  546],
       [  74, 6437]])

In [ ]:
ytest = np.array(ytest).ravel()
y_predgb= np.array(y_predgb).ravel()
df_compa = pd.DataFrame({'Nilai Sesungguhnya': ytest, 'Nilai Prediksi': y_predgb})
df_compa.head(10)

Nilai Sesungguhnya  Nilai Prediksi
0                   1               1
1                   0               1
2                   1               1
3                   1               1
4                   1               1
5                   0               1
6                   1               1
7                   1               1
8                   0               1
9                   1               1

In [ ]:
#load data test
data_test = 'https://docs.google.com/spreadsheets/d/e/2PACX-1vTDODgW9YxoP5iLyAcJTnFaMTR8_rc7ZMR2kkASoWzCmmrVpB-AsT5CVWMoUO5fC8U5Anb10oWmnOuA/pub?gid=1874408194&single=true&output=csv'
test = pd.read_csv(data_test)

test_2 = test.drop(['id','DC206','DC207','DC208','DC209','DC213','DC214','DC215','DC216','DC220','DC230a','DC230b','DC232','DC232b','DC235','DC241','DC242','DC244','DC246','DC252','DC142a'],axis=1)

#Make the predictions
predict= gb.predict(test_2)


In [ ]:
id =test['id']
predict= np.array(predict).ravel()
df_pred = pd.DataFrame({'id': id, 'DC201': predict})
df_pred['DC201'] = df_pred.DC201.map({1:'Layak Minum', 0: 'Tidak Layak Minum'})
df_pred.head(10)

id              DC201
0  26718        Layak Minum
1  26802        Layak Minum
2  41302        Layak Minum
3  38698        Layak Minum
4  44257        Layak Minum
5   4087  Tidak Layak Minum
6   2543        Layak Minum
7  39565  Tidak Layak Minum
8   4871        Layak Minum
9  46131  Tidak Layak Minum

In [ ]:
df_pred.to_csv ('Submissions_AutoWin')